docs: io annotation spec (#46)
LEQO-Framework/leqo-backend: docs/usage/annotations.rst

Reviewer: This should probably contain a full example (at the end) with a modelling block (screenshot, ascii art or similar) and the corresponding qasm code (only qubit declarations and aliases need to be spelled out, actual code can be a comment `// implementation here`) with the required annotations. (Maybe even 2 such blocks with connected inputs and outputs and how this will be transformed/stitched together <- this may be easier once you can generate such an output with the backend)

Author: See #56

Annotations
===========

A developer can create mappings between a concrete implementation (openqasm) and a program modelled in the `frontend <https://github.com/LEQO-Framework/low-code-modeler>`_ by using various openqasm3 annotations.

> Annotations can be emulated in openqasm2 by using special comments.

Reviewer 2: minor comments: modelled -> modeled (to stay consistent)

.. warning::
    The `whole line <https://openqasm.com/language/directives.html#annotations#:~:text=continue%20to%20the%20end%20of%20the%20line>`_ will be interpreted like an annotation.
    Therefore you cannot use inline-comments on annotations!

Input
-----

> One input is defined as a single qubit register (:class:`~openqasm3.ast.QubitDeclaration`) with a single `@leqo.input` annotation.

Reviewer: ```suggestion
One input is defined as a single qubit register (:class:`~openqasm3.ast.QubitDeclaration`) with a single `@leqo.input` annotation followed by the index of that input.
The input index ...
```
Format (i.e., a single positive? integer number) and **purpose** of the input index should be presented here already (at least briefly). Not everyone will read the bullet points below carefully.

* Inputs can be of arbitrary size (See :ref:`input-memory-layout`)
* Inputs have to be defined as contiguous memory
    One :class:`~openqasm3.ast.QubitDeclaration` corresponds to one input
* Input indices must be selected from a contiguous range of integers starting at `0`
   No skips, no duplicates
* Inputs may be declared anywhere in code
* Input annotations may only appear on a :class:`~openqasm3.ast.QubitDeclaration`
* Input annotations may only appear once per statement

.. code-block:: openqasm3
    :linenos:

    // Qubit array

>     @leqo.input <<InputIndex>>

Reviewer: For placeholders, it is usually a good idea to explicitly document what is part of the placeholder and how it should be replaced (i.e., replace `<<InputIndex>>` including the angled brackets with the index number of the input)

    qubit[<<length>>] someName;

    // Single qubit
    @leqo.input <<InputIndex>>
    qubit someName;

.. _input-memory-layout:

Memory Layout
~~~~~~~~~~~~~

There is no explicit limit to the size (qubit count) of an input.

> Keep in mind though that the number of physical qubits on a real device is limited.

Reviewer: ```suggestion
However, the number of physical qubits on a real device is limited.
```

See :ref:`reusable-qubit-annotation` for workarounds.

> The annotated input size must match the input size of the corresponding node from the `visual model <https://github.com/LEQO-Framework/low-code-modeler>`_.

Reviewer: Bit ordering should also be specified explicitly (even if obvious)

The backend actively ensures that input memory is initialized.

> All other qubits still have to be assumed to be in an undefined state (See `OpenQasm Specification <https://openqasm.com/language/types.html#qubits#:~:text=Qubits%20are%20initially%20in%20an%20undefined%20state>`_ and :ref:`reusable-qubit-annotation`).

Reviewer: This probably needs some more thought. We already talked about treating them as pristine (or reusable in the terms of this document) ancilla qubits (i.e. |0⟩). Most algorithms will require those.
(But there are some algorithms that can work with (some) dirty qubits, i.e., qubits in an undefined state. For inputs, dirty ancillas are the exception so we might want to add another annotation to mark those.)

Author: As per [spec](https://openqasm.com/language/types.html#qubits#:~:text=Qubits%20are%20initially%20in%20an%20undefined%20state):
> Qubits are initially in an undefined state. A quantum reset operation is one way to initialize qubit states.

The node-implementation that's emitted as the fist one of the merged program cannot get reused qubits and will get qubits in an undefined state.

So, in my opinion, the backend should not guarantee any state for any qubits and does not need to demand reusable qubits to be `|0⟩`.

Reviewer: Yes, the spec says so. But the qasm3 spec is still work in progress and adoption of the spec is very slow. (In practice, this means that the spec is often less relevant than what the QPUs / IBM actually do.)

The current de-facto default is that qubits are initialized in the `|0⟩` state before the circuit instructions are executed. Even the Qiskit Qasm3 export assumes this as default and does not emit `reset` operations for all qubits. This is one reason why we should adopt that standard/convention too.

The other reason, why the backend must guarantee predictable qubit initializations is, that a `reset` operation destroys the current qubit state. This can only be done safely if the qubit was either never used or was uncomputed beforehand. Both can only be guaranteed by the backend and not by the qasm code using the qubit!

In practice, it should suffice to require uncompute (and the reusable qubit output) to include the `reset` operation so that those qubits are guaranteed to be `|0⟩`. As the provider of the qasm for the uncompute needs to fulfill this requirement, there should be no additional work for the backend.

Author: As one of the goals of this project was to be platform-agnostic, I'd rather stick to the spec as it stands than to to some vendor-specific convention. We can of always adapt to a chaning spec.

The backend will always ensure that new (by default) or reused qubits (user has to ensure this when annotating a qubit with `@leqo.reusable`) will not be entangled.

What I'm proposing is to force the user to manually initialize the state of all qubits (by using `h`, `reset`, etc) before usage to be compliant with the spec and by extension all other valid implementations of openqasm3.
That way we don't need to require reusable qubits to be `reset`; they just need to not be entangled.

The difference is that other valid openqasm3 implementations might choose to not initialize the "new" qubits (not the reused once). The backend has no control over that and as long as we don't simulate IBM / Qiskit behavior by automatically inserting `reset` statements the resulting qasm might be broken on other platforms.

Reviewer: This is not a vendor-specific convention. Pretty much every QPU does it like this (and it is unlikely that this will change in the future). The qasm3 spec is more conservative, probably to allow for changing this convention if required, but currently this would be the one point I suggest not following the spec by the letter.

We have 2 cases:
1. initial qubits: Those will (currently) be initialized as `|0⟩` even without specific `reset` instructions.
2. reusable qubits: Must be guaranteed to be free of entanglement. Algorithms will universally need them in a defined state (i.e., the `|0⟩`) state. Depending on the uncompute operation, this may not require an actual `reset` operation.

In both cases the actual `reset` operation is optional (or at least can be optional). However, the code still assumes `|0⟩` states. The point here is that `|0⟩` states can be guaranteed without explicit `reset` statements in the circuit. Currently, it is not guaranteed that Vendors even implement such a `reset` operation. (Even when they are, they have a very high cost of execution as they usually require a measurement, which takes longer than a simple gate operation.) In summary, only the uncompute can know the most efficient way to reset the qubit to the `|0⟩` state, thus, the uncompute must include that reset! (=> case 2 MUST require `|0⟩` states for qubits)

For the remaining case 1: If there ever is a change in the default initialization, we just need to prefix all generated circuits with an extra layer of `reset` or `x` or whatever operations to get the current behavior back. This is easy to do and should be left to the future as required.

There will inevitably be inconsistencies in implementations of other vendors. Even if they all implement the same spec, they currently all have only partial implementations of an (as of yet incomplete) qasm3 spec. This includes IBM. As IBM has the most features implemented (and generally the most mature and accessible/testable implementation), this is our current development target for the actual behavior (and not the spec!). Inconsistencies in the vendor implementations can and should be dealt with by our Qunicorn project. What this means for you is: If you develop against Qiskit/IBM and use Qunicorn to execute the circuits, then you are platform-agnostic.

Reviewer 3: Thank you for that comprehensive answer. Does this means we have to differentiate between following cases?

- qubit declarations
  1. input: from another snippet
  2. ancila: are in the |0⟩ and not entangled (this is the default, if no annotation)
  3. ancila_robust: are in some state and not entangled
- qubit alias
  1. output: to another snippet
  2. reusable: |0⟩ and not entangled
  3. reusable_limited: some state and not entangled
  4. reusable_after_uncompute: |0⟩ and not entangled after uncompute was executed
  5. reusable_limited_after_uncompute: some state and not entangled after uncompute was executed
  6. trash: entangled qubits (this is the default, if no annotation)

Do you have suggestions how to call/annotate them? And we could consider removing some cases (reusable_limited_after_uncompute?) for usability.

Reviewer: It is a bit simpler that this, especially for output qubits

From the perspective of a single Block:

* input qubits
    1. input from another snippet -> requires annotation
    2. ancilla qubits -> the default, without annotation; must be |0⟩ and not entangled
    3. "dirty" ancilla qubits -> qubits in ANY state and possibly entangled, require opt in via annotation; this case is rare and could be specified later
* output qubits
    1. output to another snippet -> requires annotation
    2. directly reusable ancillas -> must always be |0⟩ and not entangled, requires annotation
    3. dirty ancilla ("trash") -> default, without annotation; a provided uncompute should turn them into reusable ancillas

The third output case is the most interesting. Those qubits can be used for snippets that allow for the third input case. Or, if an uncompute is provided, can be turned into reusable ancillas (same as the second output case). The uncompute should just be required (for now) to take exactly all the qubits of the qasm snippet (i.e. have the exact same register definitions etc.) and can mark additional qubits as reusable ancilla qubits (other input and output annotations should be ignored for uncompute snippets).

From the perspective of the entire program, we additionally have the starting qubits. Those should be assumed to be initialized in the same condition as reusable ancilla qubits.

Naming:
* input/output qubits / q(ubit)-registers (those names are ok)
* reusable ancilla qubits -> ancilla qubits / workspace qubits  (the reusable is the default and kinda long...)
* dirty ancilla / trash -> I would go with "dirty ancilla qubits" for now but there may be better (e.g., more positive) names ("used ancilla qubits", "unknown state qubits", ???)

Wikipedia "ancilla" background: <https://en.wikipedia.org/wiki/Ancilla_bit>

Reviewer: Your example is mostly correct:

* s0q1 can be used for any qubit in s1, but should probably be preferred for s1q0
* s0q2 can be used as is for s1q1, when the uncompute was applied, it can be used the same as s0q1
* s0q3 can be used as is for s1q1

An algorithm that can use a dirty ancilla (i.e., a qubit in an unknown state) can always use a reset ancilla (i.e., |0⟩). (The |0⟩ state is included in all states, thus the unknown state can also be the |0⟩ state.) This can be read as s1q1 being more flexible in what it accepts.

We do not need an annotation for "not entangled, but unknown state". Those qubits would be much more useful as reusable ancillas.

About the (possible) entanglement between s0q3 and s0q0: If s0q3 is used in a computation, this can affect the state of s0q0, which might throw off the final result. However, there can be operations that do not affect the entangled qubits. For example a Pauli X gate followed by another Pauli X gate (X*X = Identity) would leave the state untouched. Some (few) algorithms can use those dirty qubits in this way, only temporarily altering their state and restoring it right after, which does not affect entanglement. This has very limited use, which is why it is the exception and reusable ancillas are the default.

(side note: oblivious != obvious)

> In the future, it will be allowed to input less qubits than specified using the annotation.

Reviewer: This is a plan, not a certainty. "will be allowed" -> "it is planned to allow ... in the future"

In this case the backend will fill the lowest bytes with the actual input and ensure the upper bytes are initialized to zero:

    .. csv-table:: Example input register of size `7`
        :header: "0", "1", "2", "3", "4", "5", "6"

        "p[0]", "p[1]", "p[2]", "p[3]", "p[4]", "p[5]", "p[6]"

>         "p[0]", "p[1]", "0", "0", "0", "0", "0"

Reviewer 3: We should the quantum syntax here:
|0⟩

.. note::

>     The input might be split into separate non-contiguous memory blocks by the backend.

Reviewer: What is the consequence of this for the user? Why would the backend even do this?

Output
------

> One output is defines as a single alias (:class:`~openqasm3.ast.AliasStatement`) with a single `@leqo.output` annotation.

Reviewer: ```suggestion
One output is defined as a single alias (:class:`~openqasm3.ast.AliasStatement`) with a single `@leqo.output` annotation.
```

* Outputs may be concatenated from multiple non-contiguous blocks of memory.
* Output indices must be selected from a contiguous range of integers starting at `0`
    No skips, no duplicates
* Outputs may be declared anywhere in code
* Outputs may be used like any other alias
* Output annotations may only appear above a :class:`~openqasm3.ast.AliasStatement` pointing to qubits

> * Output annotations may only appear once per statement

Reviewer: And Qubits may only be used/aliased in one or no outputs (not in multiple outputs)

.. code-block:: openqasm3
    :linenos:

    @leqo.output <<OutputIndex>>
    let someOutput = <<Expression>>;

    // Example
    @leqo.output 0

>     let output1 = a[1:2:3] ++ b[{1,2,3}];

Reviewer: Maybe include definitions for a and b in the example

.. note::

>     Your defined alias might never be used in code.
>     Still, you have to define one including an identifier chosen by you.
>     The identifier is insignificant and will be ignored.

Reviewer: ```suggestion
    Even if the ouput alias is not used in the code, an alias for the output must be defined to mark the qubits that should be treated as outputs.
    The identifier is insignificant and will be ignored.
```

.. _reusable-qubit-annotation:

> Reusable Qubits

Reviewer: Maybe add "(Ancilla Qubits)" to the heading (or somewhere in the text). The term "ancilla qubits" is the commonly used term for this kind of workspace qubits and thus should be mentioned here.

-----------------

If the programmer manually resets a qubit they can mark it as reusable.
To do so, one can declare an alias to the reusable qubits.

* Reusable annotated aliases may be declared anywhere in code
* Reusable annotated aliases may be used like any other alias
* Reusable annotations may only appear above a :class:`~openqasm3.ast.AliasStatement` pointing to qubits
* Reusable annotations may only appear once per statement

> * Reusable annotations guarantee that the backend is free to reuse the qubit (i.e. it is not entangled any more)

Reviewer: It should also be in the |0⟩ state.

Reviewer 2: A qubit must be manually reset by the programmer to |0> before it can be marked as reusable. The backend assumes that the programmer has properly reset the qubit before using the @leqo.reusable annotation.

Author: See https://github.com/LEQO-Framework/leqo-backend/pull/46#discussion_r2021369819

Author: We now require initialization for reusable qubits

   No skips, no duplicates
* Inputs may be declared anywhere in code
* Input annotations may only appear on a :class:`~openqasm3.ast.QubitDeclaration`
* Input annotations may only appear once per statement

.. code-block:: openqasm3
    :linenos:

    // Qubit array
    @leqo.input <<InputIndex>>
    qubit[<<length>>] someName;

.. code-block:: openqasm3
    :linenos:

    // Single qubit
    @leqo.input <<InputIndex>>
    qubit someName;

`<<InputIndex>>` is replaced with the index of an input (positive integer literal)

.. code-block:: openqasm3
    :linenos:

    // Example
    @leqo.input 0
    qubit someName;

.. _input-memory-layout:

Memory Layout
~~~~~~~~~~~~~

There is no explicit limit to the size (qubit count) of an input.
However, the number of physical qubits on a real device is limited.
See :ref:`reusable-qubit-annotation` for workarounds.

The annotated input size must match the input size of the corresponding node from the `visual model <https://github.com/LEQO-Framework/low-code-modeler>`_.
Inputs are expected to be **Little Endian**.

The backend actively ensures that input memory is initialized.

> All other qubits still have to be assumed to be in an undefined state (See `OpenQasm Specification <https://openqasm.com/language/types.html#qubits#:~:text=Qubits%20are%20initially%20in%20an%20undefined%20state>`_ and :ref:`reusable-qubit-annotation`).

Author: I'd like to do this in this PR.
Do you happen to have a source I could link for IBM initialization behavior?

Annotations
===========

A developer can create mappings between a concrete implementation (openqasm) and a program modelled in the frontend by using various openqasm3 annotations.

> Annotations can be emulated in openqasm2 by using special comments.

Reviewer 3: We will need an example of that

Author: I'd link to a separate page, once #22 is merged as we wanted to implement a generic concept to convert openqasm2 comments to openqasm3 annotations

Reviewer 3: Ok, sounds like a good idea.

Input
-----

> One input is defined as a single :class:`~openqasm3.ast.QubitDeclaration` with a single `@leqo.input` annotation.

Reviewer 3: We should not assume that the user has knowledge about the python module openqasm3.ast.
Instead, we should use the the name for this used in the openqasm3 language.

Author: This will generate a link with the content `QubitDeclaration` in the final page.
We could of course just write `qubit declaration`

Reviewer 3: I would prefer to use the name used in https://openqasm.com/language/types.html#quantum-types, but this seems also fine.

> * Inputs can be of arbitrary size

Reviewer 3: We should add a comment, that this size has to map to the size used in the model, but may be less (but not more).

Author: Good point!

Reviewer 3: Should I do that in a own commit, or do you want to do it?

* Inputs need to be defined as contiguous memory
    One :class:`~openqasm3.ast.QubitDeclaration` corresponds to one input
* Input indices must be selected from a contiguous range of integers starting at `0`
   No skips, no duplicates
* Inputs may be declared anywhere in code
* Input annotations may only appear on a :class:`~openqasm3.ast.QubitDeclaration`
* Input annotations may only appear once per statement

.. code-block:: openqasm3
    :linenos:

    // Qubit array
    @leqo.input <<InputIndex>>
    qubit[<<length>>] someName;

    // Single qubit
    @leqo.input <<InputIndex>>
    qubit someName;

.. note::

>     The input might be split into separate non-contiguous memory blocks by the processor.

Reviewer 3: You mean by the compiler?

Author: Yes. As this is an implementation detail, we could move it to the "internal" section of the docs.

Output
------

One output is defines as a single :class:`~openqasm3.ast.AliasStatement` with a single `@leqo.output` annotation.

* Outputs may be concatenated from multiple non-contiguous blocks of memory.
* Output indices must be selected from a contiguous range of integers starting at `0`
    No skips, no duplicates
* Outputs may be declared anywhere in code

> * Output annotations may only appear above a :class:`~openqasm3.ast.AliasStatement` pointing to qubits

Reviewer 3: Same here, just use "alias" statement

* Output annotations may only appear once per statement

.. code-block:: openqasm3
    :linenos:

    @leqo.output <<OutputIndex>>
    let someOutput = <<Expression>>;

    // Example
    @leqo.output 0
    let output1 = a[1:2:3] ++ b[{1,2,3}];

Reusable Qubits
-----------------

If the programmer manually resets a qubit they can mark it as reusable.
To do so, one can declare an alias to the reusable qubits.

* Reusable annotations may only appear above a :class:`~openqasm3.ast.AliasStatement` pointing to qubits
* Reusable annotations may only appear once per statement

.. code-block:: openqasm3
    :linenos:

    @leqo.reusable

>     let reusable1 = <<Expression>>;

Reviewer 3: We should note, that the name used for the alias does not matter.

Author: Same for `@leqo.input` and `@leqo.output`.
As we did not state they are important we might as well don't mention the opposite...

Reviewer 3: I think this would be good to note here (and maybe for the output as well), because the aliases will likely be not used by the user, and he will wonder how to call them.
